core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -298,45 +351,46 @@ import com.datastax.driver.core.utils.UUIDs
+          updateStageStage(_.copy(fromOffset = repr.offset)
+            .tagPidSequenceNumberUpdate(repr.persistenceId, (1, repr.offset)))
           push(out, repr)
           false
-        } else if (usingOffset && (currTimeBucket.inPast || settings.eventsByTagNewPersistenceIdScanTimeout == Duration.Zero)) {
+        } else if (usingOffset && (stageState.currentTimeBucket.inPast || settings.eventsByTagNewPersistenceIdScanTimeout == Duration.Zero)) {
           // If we're in the past and this is an offset query we assume this is
           // the first tagPidSequenceNr
-          log.debug("New persistence id: {}. Timebucket: {}. Tag pid sequence nr: {}", repr.persistenceId, currTimeBucket, repr.tagPidSequenceNr)
-          currentOffset = repr.offset
-          tagPidSequenceNrs += (repr.persistenceId -> ((repr.tagPidSequenceNr, repr.offset)))
+          log.debug(
+            "[{}] New persistence id: {}. Timebucket: {}. Tag pid sequence nr: {}",
+            stageUuid, repr.persistenceId, stageState.currentTimeBucket, repr.tagPidSequenceNr)
+          updateStageStage(_.copy(fromOffset = repr.offset)
+            .tagPidSequenceNumberUpdate(repr.persistenceId, (repr.tagPidSequenceNr, repr.offset)))
           push(out, repr)
           false
         } else {
           log.info(
-            "{}: Missing event for new persistence id: {}. Expected sequence nr: {}, actual: {}.",
-            session.tag, repr.persistenceId, expectedSequenceNr, repr.tagPidSequenceNr
-          )
-          val previousBucketStart = UUIDs.startOf(currTimeBucket.previous(1).key)
+            s"[${stageUuid}] " + " {}: Missing event for new persistence id: {}. Expected tag pid sequence nr: {}, actual: {}.",
+            session.tag, repr.persistenceId, expectedSequenceNr, repr.tagPidSequenceNr)
+          val previousBucketStart = UUIDs.startOf(stageState.currentTimeBucket.previous(1).key)
           val startingOffset: UUID = if (UUIDComparator.comparator.compare(previousBucketStart, fromOffset) < 0) {
-            log.debug("Starting at fromOffset")
+            log.debug("[{}] Starting at fromOffset", stageUuid)
             fromOffset
           } else {
-            log.debug("Starting at startOfBucket")
+            log.debug("[{}] Starting at startOfBucket", stageUuid)
             previousBucketStart
           }
           val bucket = TimeBucket(startingOffset, bucketSize)
           val lookInPrevious = !bucket.within(startingOffset)
-          missingLookup = Some(LookingForMissing(
-            repr :: Nil,
-            startingOffset,
-            bucket,
-            lookInPrevious,
-            repr.offset,
-            repr.persistenceId,
-            repr.tagPidSequenceNr,
-            (1L until repr.tagPidSequenceNr).toSet,
-            failIfNotFound = false,
-            deadline = Deadline.now + settings.eventsByTagNewPersistenceIdScanTimeout
-          ))
+          updateStageStage(_.copy(missingLookup =
```

Comment:
updateStageStage?

should it be updateStageState?

Reply:
Yep! The multiple vars were starting to get hard to track so I added this, not sure it is any clearer